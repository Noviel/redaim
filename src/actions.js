import { TRIGGER_UNKNOWN, TARGET_UNKNOWN, EXEC_UNKNOWN } from './constants';

export const createAction = 
  (trigger = TRIGGER_UNKNOWN) => 
  (target = TARGET_UNKNOWN) => 
  (action = {}) => {
  
  const { exec = EXEC_UNKNOWN, payload = null, meta = {}, ...rest } = action;

  return { 
    type: trigger, 
    meta: { target, exec, ...meta },
    payload,
    ...rest
  };
};

export const createActionReversed = 
  (action = {}) => 
  (target = TARGET_UNKNOWN) => 
  (trigger = TRIGGER_UNKNOWN) => {
  
  const { exec = EXEC_UNKNOWN, payload = null, meta = {}, ...rest } = action;

  return { 
    type: trigger, 
    meta: { target, exec, ...meta },
    payload,
    ...rest
  };
};

export const wrapActions = fn => actions => {
  const finalActions = {};
  
  for (const a in actions) {
    const currAction = actions[a];
    if (typeof currAction === 'function') {
      finalActions[a] = (...args) => fn(currAction(...args));
    } else {
      finalActions[a] = (...args) => fn(currAction, ...args);
    }
  }

  return finalActions;
};

export const bindActions = actionCreatorWithTrigger => target => actions =>
  wrapActions(actionCreatorWithTrigger(target))(actions);

export const createActions = trigger => {
  const boundToTrigger = bindActions(createAction(trigger));

  return target => { 
    const boundToTarget = boundToTrigger(target);

    return actions => boundToTarget(actions);
  };
};
