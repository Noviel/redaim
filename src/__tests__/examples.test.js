import { createAction, createActions } from '../actions';

describe('create actions examples', () => {
  it('should return correct action', () => {
    const action = createAction('COUNTERS_TRIGGER')('counter1')({
      exec: 'ADD',
      meta: { additionalMetaField: 42 },
      payload: { value: 100 }
    });

    expect(action).toMatchSnapshot();
  });

  it('should create actions from action object with functions as keys', () => {
    const actionsObject = {
      add: { exec: 'ADD', payload: { value: 100 } },
      sub: value => ({ exec: 'SUB', payload: { value } })
    };

    const { add, sub } = createActions('COUNTER_TRIGGER')('counter1')(actionsObject);

    const result = {
      add: add(),
      sub: sub(10)
    };

    expect(result).toMatchSnapshot();
  });
});
