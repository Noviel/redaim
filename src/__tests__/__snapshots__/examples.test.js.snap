// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create actions examples should create actions from action object with functions as keys 1`] = `
Object {
  "add": Object {
    "meta": Object {
      "exec": "ADD",
      "target": "counter1",
    },
    "payload": Object {
      "value": 100,
    },
    "type": "COUNTER_TRIGGER",
  },
  "sub": Object {
    "meta": Object {
      "exec": "SUB",
      "target": "counter1",
    },
    "payload": Object {
      "value": 10,
    },
    "type": "COUNTER_TRIGGER",
  },
}
`;

exports[`create actions examples should return correct action 1`] = `
Object {
  "meta": Object {
    "additionalMetaField": 42,
    "exec": "ADD",
    "target": "counter1",
  },
  "payload": Object {
    "value": 100,
  },
  "type": "COUNTERS_TRIGGER",
}
`;
